scripts/dts/edtsdatabase.py

```diff
@@ -0,0 +1,193 @@
+#   'compatibles':  dict(compatible : sorted list(device-id)),
+#   'device-types': dict(device-type : sorted list(device-id)),
+#   ...
+# )
+#
+# device-struct: dict(
+#   'device-id' : device-id,
+#   'device-type' : list(device-type) or device-type,
+#   'compatible' : list(compatible) or compatible,
+#   'label' : label,
+#   property-name : property-value ...
+# )
+#
+# Database types:
+#
+# device-id: <compatible>_<base address in hex>_<offset in hex>,
+# compatible: any of ['st,stm32-spi-fifo', ...] - 'compatibe' from <binding>.yaml
+# device-type: any of ['GPIO', 'SPI', 'CAN', ...] - 'id' from <binding>.yaml
+# label: any of ['UART_0', 'SPI_11', ...] - label directive from DTS
+#
+class EDTSDatabase(Mapping):
+
+    def __init__(self, *args, **kw):
+        self._edts = dict(*args, **kw)
+        # setup basic database schema
+        for edts_key in ('devices', 'compatibles', 'device-types'):
+            if not edts_key in self._edts:
+                self._edts[edts_key] = dict()
+
+    def __getitem__(self, key):
+        return self._edts[key]
+
+    def __iter__(self):
+        return iter(self._edts)
+
+    def __len__(self):
+        return len(self._edts)
+
+    def convert_string_to_label(self, s):
+        # Transmute ,-@/ to _
+        s = s.replace("-", "_")
+        s = s.replace(",", "_")
+        s = s.replace("@", "_")
+        s = s.replace("/", "_")
+        # Uppercase the string
+        s = s.upper()
+        return s
+
+    def device_id(self, compatible, base_address, offset):
+        # sanitize
+        base_address = int(base_address)
+        offset = int(offset)
+        compatibe = self.convert_string_to_label(compatible.strip())
+        return "{}_{:08X}_{:02X}".format(compatible, base_address, offset)
+
+    def device_id_by_label(self, label):
+        for device in self._edts['devices']:
+            if label == device['label']:
+                return device['device-id']
+
+    def _update_device_compatible(self, device_id, compatible):
+        if compatible not in self._edts['compatibles']:
+            self._edts['compatibles'][compatible] = list()
+        if device_id not in self._edts['compatibles'][compatible]:
+            self._edts['compatibles'][compatible].append(device_id)
+            self._edts['compatibles'][compatible].sort()
+
+    def _update_device_type(self, device_id, device_type):
+        if device_type not in self._edts['device-types']:
+            self._edts['device-types'][device_type] = list()
+        if device_id not in self._edts['device-types'][device_type]:
+            self._edts['device-types'][device_type].append(device_id)
+            self._edts['device-types'][device_type].sort()
+
+    ##
+    # @brief Insert property value for the device of the given device id.
+    #
+    # @param device_id
+    # @param property_path Path of the property to access
+    #                      (e.g. 'reg/0', 'interrupts/prio', 'label', ...)
+    # @param property_value value
```

Comment:
The way it works today is that we enter a top level property and a dict associated.
For instance:
'deflabel: interrupts'
("new_structs: [{'data': [33, 0], 'labels': ['irq', 'priority']}, {'data': "
 "[34, 0], 'labels': ['irq', 'priority']}]")
''
Would be easier to stick to that in a first iteration

Reply:
This would become:

```
device_id = edts.device_id(node_address, ....) # or other helper function - TBD
edts.device_insert_property(device_id, 'interrupts', [{'data': [33, 0], 'labels': ['irq', 'priority']}, {'data': "
"[34, 0], 'labels': ['irq', 'priority']}])
```